[Win32 Signals] Add term and ctrl-c signal handlers

davinci26 opened the pull request:
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Part 1 of https://github.com/envoyproxy/envoy/issues/13188. Adds support for `ctrl+c` and `ctrl+break` for Envoy on Windows.

The implementation for this following:

1. On `platform_impl`  we register a `CtrlHandler` which runs on a separate thread.
2. On `signal_impl` we register a read event reader.

Thread (1) and (2) communicate via a socket pair and the event is handled on Windows the same way as it is handled on POSIX

Risk Level: Low (Windows Only)
Testing: Manual testing + integration tests
Docs Changes: N/A
Release Notes: N/A

davinci26 commented:
@envoyproxy/windows-dev

wrowe commented on include/envoy/common/platform.h:
```diff
@@ -151,6 +152,11 @@ struct msghdr {
+#define ENVOY_SIGINIT CTRL_C_EVENT
+#define ENVOY_SIGTERM CTRL_CLOSE_EVENT
+#define ENVOY_SIGUSR1 -1
+#define ENVOY_SIGHUP -1
```
It seems that CTRL_LOGOFF_EVENT and CTRL_SHUTDOWN_EVENT must follow the ENVOY_SIGTERM path as any linux would cause during shutdown, or envoy will have no chance to exit cleanly but os shutdown will be delayed.

The CTRL_BREAK_EVENT is really interesting; this could be mapped to graceful shutdown or hot restart.

sunjayBhatia replied:
(Also CTRL SHUTDOWN is quite relevant for running Envoy via Windows containers as I believe that is the only notification/event a process will be able to handle when a container is being destroyed. Container orchestrators may also choose to do some lifecycle management with the Admin API `/quitquitquit` endpoint instead, but handling CTRL SHUTDOWN would cover other cases.)

antoniovicente replied:
Is your suggestion to register some additional signal events so they also go down the clean shutdown path?  Perhaps we should add a function that returns a list of signals that should trigger the clean shutdown handler

Should config for fast vs clean shutdown on CTRL_LOGOFF_EVENT and CTRL_BREAK_EVENT make sense?  I could see fast teardown being desirable on logoff.  Clean shutdown on both of these events may be most appropriate.

davinci26 replied:
All the signals in the control handler map to clean shutdown so I just created an event `ENVOY_WIN32_SIGTERM` and all the `CtrlHandler` events are mapped to it. 

All the Windows events are causing a clean shutdown so we dont need a function for now. When we add support for SCM if might have other events that are not related to shutdown and at that point we might need to implement a function

davinci26 replied:
Is there a concept of clean vs fast shutdown?

wrowe reviewed (changes requested):
This approach is otherwise looking good to me. In the shutdown/logoff cases, I've found that returning immediately from the CtrlHandler causes the typical "now attempt to kill" reaction immediately from the OS, so a delay is usually helpful.

sunjayBhatia commented on test/exe/outofproc_main_test.py:
```diff
@@ -0,0 +1,74 @@
+    self.envoy_proc.send_signal(ctrl_break_signal)
+    time.sleep(2)
+    poll = self.envoy_proc.poll()
+    self.assertTrue(poll != None)
```
The assertion that we wait for a bit and the process has exited seems a little loose, if we were to remove the CTRL-BREAK handling logic wouldn't this test still pass?

Could we do some more assertions around maybe synchronizing the process exit or checking for some log line or stat or something that indicates graceful shutdown was invoked via signal? Some similar tests use a test main (instead of the real one) to do things like this, since we have log lines already in the dispatcher callback this might be doable without that?

wrowe replied:
> The assertion that we wait for a bit and the process has exited seems a little loose

Agreed, we should determine our process has finished and then return, not some "pause 30 seconds" hack.

My understanding, ctrl+break/ctrl+close/ctrl+c do not have to be honored, so the timing of the return from those ctrl requests doesn't matter.

Shutdown/logout is time sensitive and manditory, so a process kill follows the return from the ctrl event. This behavior may have evolved since I last implemented it.

davinci26 replied:
> Some similar tests use a test main (instead of the real one) to do things like this, since we have log lines already in the dispatcher callback this might be doable without that?

I can parse the standard output to verify that the event was indeed caught and processed.

davinci26 replied:
In terms of testing I did the following:

1. I added a "hacky" sleep statement that is responsible for waiting while Envoy is constructed. The alternative to that would be to keep pinging the Envoy ip/port and see when it responds back to know that it is fully initialized.
2. Wait for Envoy to close after we send the singal.
3. Read the standard output to verify that we closed with `ENVOY_WIN32_SIGTERM`

The test needs to be executed outside of bazel as a standalone executable because of the way the console signals work. Bazel is handling them for us and as a result the envoy process never sees them.

sunjayBhatia commented on .bazelrc:
```diff
@@ -302,6 +302,7 @@ build:windows --action_env=TMPDIR
 build:windows --define signal_trace=disabled
 build:windows --define hot_restart=disabled
 build:windows --define tcmalloc=disabled
+build:windows --define wasm=disabled
```
Could you also remove this from `ci/windows_ci_steps.sh` just to clean this up

davinci26 replied:
done

sunjayBhatia commented on source/server/server.cc:
```diff
@@ -614,28 +617,29 @@ RunHelper::RunHelper(Instance& instance, const Options& options, Event::Dispatch
-      ENVOY_LOG(warn, "caught and eating SIGHUP. See documentation for how to hot restart.");
-    });
-#endif
+    if constexpr (platformSupportsSignal(ENVOY_SIGTERM)) {
```
Maybe out of scope of this PR and worth doing in a follow up, but it seems like this block could be abstracted out of this file at least, could enable some more detailed test coverage with mocking etc.

antoniovicente commented on source/server/server.cc:
```diff
@@ -614,28 +617,29 @@ RunHelper::RunHelper(Instance& instance, const Options& options, Event::Dispatch
+        ENVOY_LOG(warn, "caught SIGINT");
+        instance.shutdown();
+      });
+    }
```
See suggestion for platform functions that return a list of signals to associate with the clean shutdown handler as a possible alternative to ENVOY_SIGTERM and friends.

https://github.com/envoyproxy/envoy/pull/13954#discussion_r521584038

davinci26 replied:
Agreed if the relationship is one-to-many then functions are the way to go. SIGTERM would map to multiple Windows events

antoniovicente reviewed (commented):
Thanks for the continued help to improve cross platform support.

davinci26 commented:
I am adding a /wait tag as I am working more with the automated test. @sunjayBhatia was right and the ctrl+break is caught from the test regardless of the signal handling. I am working on adding ctrl + c signals in the tests but it is a bit tricky.

davinci26 commented:
/wait

antoniovicente commented:
Let me know when it is ready for another round of review.

davinci26 commented on source/server/server.cc:
```diff
@@ -53,6 +53,9 @@
 #include "server/ssl_context_manager.h"
 
+namespace {
+constexpr bool platformSupportsSignal(signal_t signal_num) { return signal_num != -1; }
```
unused

davinci26 replied:
fixed in c484a7512620d167f4f04d6d13568e3714200ea5

davinci26 commented on source/server/server.cc:
```diff
@@ -614,25 +617,28 @@ RunHelper::RunHelper(Instance& instance, const Options& options, Event::Dispatch
+      instance.shutdown();
+    });
+#else
+    sigterm_ = dispatcher.listenForSignal(ENVOY_SIGTERM, [&instance]() {
```
revert POSIX changes

davinci26 replied:
fixed in c484a7512620d167f4f04d6d13568e3714200ea5

sunjayBhatia commented on source/server/server.cc:
```diff
@@ -614,15 +614,18 @@ RunHelper::RunHelper(Instance& instance, const Options& options, Event::Dispatch
     });
 
-    sigint_ = dispatcher.listenForSignal(SIGINT, [&instance]() {
+    sigint_ = dispatcher.listenForSignal(SIGINIT, [&instance]() {
```
nit: typo should be `SIGINT` (no change)

sunjayBhatia replied:
(looks like that is a compile error anyway)

davinci26 replied:
I saw that but I am on a different VM so I am waiting for a tiny bit to fix it.

davinci26 replied:
fixed

davinci26 commented:
I expect the CI to go green now ( maybe with the exception of `clang-tidy`) @wrowe @antoniovicente ready for another review.

antoniovicente commented on source/common/event/BUILD:
```diff
@@ -13,15 +16,24 @@ envoy_cc_library(
+        "//conditions:default": [
+            "posix/signal_impl.h",
+        ],
+    }),
```
This header is provided by :signal_lib, could we remove it from the dispatcher library?

davinci26 replied:
I run into https://github.com/bazelbuild/bazel/issues/6337 if I remove it. This is why I added it

antoniovicente replied:
Ah, ok.

antoniovicente commented on source/common/event/BUILD:
```diff
@@ -40,6 +52,36 @@ envoy_cc_library(
+    name = "signal_impl_lib",
+    srcs = ["posix/signal_impl.cc"],
+    hdrs = ["posix/signal_impl.h"],
+    strip_include_prefix = "posix",
```
the strip_include_prefix may be a source of confusion, but there is prior art in source/common/common/BUILD and source/common/filesystem/BUILD so I guess this is fine.

I would have suggested having a signal_impl.h that includes the right platform header.

antoniovicente commented on source/exe/BUILD:
```diff
@@ -170,6 +170,7 @@ envoy_cc_win32_library(
         ":platform_header_lib",
         "//source/common/common:assert_lib",
         "//source/common/common:thread_lib",
+        "//source/common/event:dispatcher_lib",
```
Signal lib, not dispatcher lib

davinci26 replied:
thanks good catch! fixed

antoniovicente commented on test/exe/BUILD:
```diff
@@ -81,3 +81,24 @@ envoy_cc_test(
+# this test cannot be executed through bazel and it must
+# be executed as a standalone executable.
+# e.g.: `./bazel-bin/test/exe/win32_outofproc_main_test.exe`
+envoy_cc_test(
```
I assume there's no explicit way to generate this test binary target only under windows.

davinci26 replied:
Based on bazel documentation 

>The manual tag keyword is treated specially. It marks the test_suite target as "manual" so that it will be ignored by the wildcard expansion and automated testing facilities.

So it should not have any negative side effect on the POSIX compilations units and it should be ignored. Also given that this test has the same deps as the other units in the compilation unit if you tried to compile the entire //test/exe build file it should be minimal overhead.

antoniovicente commented on include/envoy/common/platform.h:
```diff
@@ -65,6 +65,7 @@ typedef uint32_t mode_t;
 
 typedef SOCKET os_fd_t;
 typedef HANDLE filesystem_os_id_t; // NOLINT(modernize-use-using)
+typedef DWORD signal_t;            // NOLINT(modernize-use-using)
```
Could you update the issue title to something like:

win32 signal: add term and ctrl-c signal handlers

davinci26 replied:
You mean the PR title right? The corresponding issue is a bit more broad because we also want to enable SCM (service control manager) which is the second way to manage a process in Windows (by running it as a service)

antoniovicente replied:
Yeah, sorry.  I meant PR title.

davinci26 commented:
/retest

sunjayBhatia commented:
I think you can just add to this: https://github.com/envoyproxy/envoy/blob/af17f954653afcab6a189673e3500b2e988f1aef/ci/run_clang_tidy.sh#L31-L35

There is a TODO to correctly run clang-tidy on Windows and MacOS

sunjayBhatia commented on ci/run_clang_tidy.sh:
```diff
@@ -31,7 +31,7 @@ echo "Generating compilation database..."
 # TODO(scw00): We should run clang-tidy against win32 impl once we have clang-cl support for Windows
 function exclude_win32_impl() {
-  grep -v source/common/filesystem/win32/ | grep -v source/common/common/win32 | grep -v source/exe/win32 | grep -v source/common/api/win32
+  grep -v source/common/filesystem/win32/ | grep -v source/common/common/win32 | grep -v source/exe/win32 | grep -v source/common/api/win32 | grep -v source/source/common/event/win32
```
I think you can leave out the leading `source/` on the addition to this line

davinci26 replied:
thanks!

davinci26 commented:
@sunjayBhatia thanks for the suggestion! It worked.

davinci26 commented:
/retest

davinci26 commented:
@wrowe and @antoniovicente PTAL when you get some time

antoniovicente commented on source/common/event/win32/signal_impl.cc:
```diff
@@ -0,0 +1,40 @@
+        cb_();
+      },
+      Event::FileTriggerType::Level, Event::FileReadyType::Read);
+  eventBridgeHandlersSingleton::get().insert({signal_num, write_handle});
```
What guarantees that this insert into the flat_hash_map ends up being thread safe?

Consider implementing this as an array instead of a map since the number of possible signal numbers is small.

davinci26 replied:
I assume the race condition that you refer to is someone hitting ctrl+c while the server main is adding new `signal_impl`. I transformed this into an array where each signal is assigned to a fixed position based on the signal id. Let me know what you think.

antoniovicente commented on source/common/event/win32/signal_impl.cc:
```diff
@@ -0,0 +1,40 @@
+  ASSERT(result.rc_ == 0);
+  auto write_handle = std::make_shared<Network::IoSocketHandleImpl>(socks[1], false, AF_INET);
+  result = write_handle->setBlocking(false);
+  ASSERT(result.rc_ == 0);
```
possible future refactor: I wished that a socket pair wakeup mechanism existed as an primitive under source/common/event .  There's several scenarios where it could prove useful, including replacing the wakeup mechanism currently used for Dispatcher::post callbacks.

davinci26 replied:
I assume the wake up mechanism is: `post_cb_->scheduleCallbackCurrentIteration();`. Makes sense, I would like to get #10871 first so we can have `AF_UNIX` socketpairs on windows and then I can add that.

I will open an issue for that so that we don't forget it in the meantime

antoniovicente replied:
It is.  Schedulable callbacks are implemented on top of libevent timers which require taking a lock on the event loop in order to trigger activation.

Wakeup via a socket pair feels more elegant.

davinci26 replied:
see #14244

antoniovicente commented on source/exe/win32/platform_impl.cc:
```diff
@@ -1,17 +1,56 @@
+  Buffer::OwnedImpl buffer;
+  constexpr absl::string_view data{"a"};
+  buffer.add(data);
+  auto result = handler_it->second->write(buffer);
```
A possible question: Why use a socket pair for wakeup instead of a higher level construct like a dispatcher post callback?  I think the answer is related to the set of actions that are allowed from a signal handler; writing to a socket is fine, taking locks is not and allocating memory is not fine.

Worth a comment?

davinci26 replied:
I added a comment.

 Initially my idea was that I did not want to have a global reference to the dispatcher. Seemed like a recipe for future disaster. Sadly this function does not allow for any additional context so communicating with a socket pair seems like the method that could cause the least amount of trouble. Also allocating memory or taking locks in this thread is risky.

antoniovicente replied:
Right.  At least the current implementation of Dispatcher post callbacks ends up taking locks on the libevent scheduler and the dispatcher.  Also, it allocates memory.

antoniovicente replied:
Also, I think you have some local changes that you have not pushed out yet.  I don't see the comments that you added.

davinci26 replied:
Fixed now in c378f5f6189624b029a4b5c17e677a25add84ce0

I was compiling the whole thing last night and did not get a chance to have a full compilation and tests so I just answered the comments, and let the code change for the next day. Sorry for the confusion

antoniovicente commented on source/common/event/win32/signal_impl.h:
```diff
@@ -0,0 +1,35 @@
+// Here we have a map from signal types to IoHandle. When we write to this handle we trigger an
+// event that notifies Envoy to act on the signal.
+using eventBridgeHandlersSingleton =
+    ThreadSafeSingleton<absl::flat_hash_map<signal_t, std::shared_ptr<Network::IoHandle>>>;
```
Is there a mechanism that clears ThreadSafeSingleton between tests?

davinci26 replied:
No it doesn't but I don't think it matters anyway. There are two reasons for that:

1. We only add new events. This means that the `map`/`array` is bounded in size.
2. There is no way to trigger these signals when you are running inside bazel or even gtest.

The only way to actually go through this code is by spawning envoy out of process. This is because the signals that we are catching are part of the console actually not signals to the process. And both bazel and gtest handle these console signals themselves so the code here is never executed.

antoniovicente commented on test/exe/win32_outofproc_main_test.cc:
```diff
@@ -0,0 +1,92 @@
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(MainCommonTest, Envoy1) {
```
nit: What's the 1 in the test name about?

davinci26 replied:
fixed in c378f5f6189624b029a4b5c17e677a25add84ce0

antoniovicente reviewed (commented):
Change looks pretty much ready to merge.  Just a few minor questions.  Sorry for the delays in review, this one really went to the back burner.

davinci26 commented:
/retest

antoniovicente reviewed (approved):
Thanks!

davinci26 commented:
@envoyproxy/senior-maintainers can I also get a review from one of you!

ggreenway commented on source/common/event/win32/signal_impl.cc:
```diff
@@ -0,0 +1,45 @@
+    : cb_(cb) {
+
+  if (signal_num > eventBridgeHandlersSingleton::get().size()) {
+    ENVOY_BUG(false, "Attempting to create SignalEventImpl with a signal id that exceeds the "
```
Should this be a PANIC/RELEASE_ASSERT? Isn't this a strict programming error?

davinci26 replied:
My train of though here is the following:

Assume that this is actually a bug that is introduced and this `ENVOY_BUG` would hit. As a user of Envoy do I prefer to have reduced signal handling i.e. I can not close envoy from `ctrl+c` or do I want Envoy to fail to start. 

If I were the user I would just prefer the former because I can still work around it and have Envoy working in my server. I would have a bad experience for sure but I wouldn't be blocked.

Again no strong opinion and happy to change it to `PANIC` just stating were I am coming from.

antoniovicente replied:
It does seem like a programming error.

davinci26 replied:
Fair. Changed it into a panic in this iteration.

ggreenway commented on source/exe/win32/platform_impl.cc:
```diff
@@ -1,17 +1,60 @@
 
 namespace Envoy {
 
+static volatile bool shutdown_pending = false;
```
On posix this is supposed to be `sig_atomic_t` I think. Is that available on windows? `volatile` seems like the wrong qualifier here.

davinci26 replied:
looking into it.

davinci26 replied:
fixed by making it an atomic

ggreenway commented on .bazelrc:
```diff
@@ -298,6 +298,7 @@ build:windows --action_env=TMPDIR
 build:windows --define signal_trace=disabled
 build:windows --define hot_restart=disabled
 build:windows --define tcmalloc=disabled
+build:windows --define wasm=disabled
```
I don't object to this change, but it seems unrelated to the main topic of this PR. Can it be in a separate PR?

davinci26 replied:
The reason it is baked into this PR is because you can't build `//source/exe:envoy-static` without this definition so it is part of my inner dev loop.

No strong opinions though.

ggreenway replied:
It's probably fine; I don't feel strongly about it.

ggreenway commented on source/exe/win32/platform_impl.cc:
```diff
@@ -1,17 +1,60 @@
+    return 0;
+  }
+
+  // This code is executed as part of a thread running under a Windows
```
What do you mean by `Windows context`?

davinci26 replied:
This is one of the callbacks that are executed from a thread created and managed by the console host and not by Envoy. I bet `Windows context` is not clear enough, I will update it

davinci26 replied:
fixed in latest commit.

ggreenway commented on source/exe/win32/platform_impl.cc:
```diff
@@ -1,17 +1,60 @@
+  // to wake up the signal handler.
+  Buffer::OwnedImpl buffer;
+  constexpr absl::string_view data{"a"};
+  buffer.add(data);
```
This allocates memory, contrary to the above comment. I think you can add an UnownedSlice to the buffer, and put the slice on the stack.

davinci26 replied:
sure will do

davinci26 replied:
I dont see an overload of `add` that takes an `UnownedSlice` as input. There is `addBufferFragment` but that allocates the slice.

Should I just add an overload or a function `addUnownedSlice`?

davinci26 replied:
Honestly I am not sure if this 1 byte allocation is going to save us if we are that close to OOM

antoniovicente replied:
I missed this issue.  The allocation of a buffer object ends up allocating memory too.  I would suggest use of writev instead of write from a buffer.

My understanding of the no-allocation rule in signal handlers is somewhat crude.  I think that the issue is that the allocation process itself takes locks; if the signal handler triggers while the thread is holding the allocator mutex the signal handler deadlocks.

davinci26 replied:
I changed it into `writev` and I opened another issue for some other windows threads that call `write` instead of `writev`

ggreenway commented on source/server/server.cc:
```diff
@@ -644,9 +644,12 @@ RunHelper::RunHelper(Instance& instance, const Options& options, Event::Dispatch
-// on Windows.
-#ifndef WIN32
+#ifdef WIN32
+    sigterm_ = dispatcher.listenForSignal(ENVOY_WIN32_SIGTERM, [&instance]() {
```
Can we make the SIGTERM handler for WIN32 and non-WIN32 the same? Maybe define `ENVOY_SIGTERM` on all platforms and use that?

davinci26 replied:
Will do that

davinci26 replied:
fixed

davinci26 commented:
@ggreenway I addressed most of the comments and I am looking for some further clarification is some comments that I have left open. Thanks for the code review!

ggreenway reviewed (approved):
LGTM
